lib/client-api/src/story_store.ts

```diff
@@ -383,8 +383,32 @@ export default class StoryStore {
     const loaders = [...this._globalMetadata.loaders, ...kindMetadata.loaders, ...storyLoaders];
 
     const finalStoryFn = (context: StoryContext) => {
-      const { passArgsFirst = true } = context.parameters;
-      return passArgsFirst ? (original as ArgsStoryFn)(context.args, context) : original(context);
+      const { args, argTypes, parameters } = context;
+      if (
+        argTypes &&
+        Object.values(argTypes).some(({ control }) => {
+          if (!control?.options) return false;
+          return Object.values(control.options).some((v: any) => v && typeof v === 'object');
+        })
+      ) {
+        once.warn(
+          'Objects cannot be used as values in control options. Use a `mapping` instead.\n\nMore info: https://storybook.js.org/docs/react/writing-stories/args#mapping-to-complex-arg-values'
+        );
+      }
```

Comment:
Hmm. Technically I don't think the story store should be looking at `argTypes[x].control`. Can we put this warning in the controls or docs addon somehow?

Reply:
Makes sense. I wasn't sure where to put this anyway.

Reply:
Moved to `ControlsPanel` in addon-controls.